### PR TITLE
Provide access to GLFW window in plugins

### PR DESCRIPTION
**Pull request by stuartmorgan**
Plugins may need to be able to access functions affecting the GLFW
window (e.g., a plugin to resize the window). This restructures the API
to create a distinction at both the C and C++ level between the window
controller, which provides access to high-level behaviors driving the
Flutter application, and the window, which provides access to functions
to affect the UI state of the window (i.e., wrapped GLFWwindow
functions).

Also provides a PluginRegistrar extension for plugins that need access
to GLFW-specific functionality.

### CODE REVIEW

File: shell/platform/glfw/client_wrapper/flutter_window_controller.cc [hunk 1]
```diff
@@ -44,45 +44,35 @@ bool FlutterWindowController::CreateWindow(
       [](const std::string& arg) -> const char* { return arg.c_str(); });
   size_t arg_count = engine_arguments.size();
 
-  window_ = FlutterDesktopCreateWindow(
+  controller_ = FlutterDesktopCreateWindow(
```
Comment (chinmaygarde) [inline shell/platform/glfw/client_wrapper/flutter_window_controller.cc]: How is this controller collected? Consider using a smart pointer for such objects (unique_ptr with a custom deleter would work fine).

Comment (chinmaygarde) [reply shell/platform/glfw/client_wrapper/flutter_window_controller.cc]: Nevermind. Saw the comment about the transfer in later.

File: shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  // The pixel format is 32-bit RGBA. The provided image data only needs to be
+  // valid for the duration of the call to this method. Pass a nullptr to revert
+  // to the default icon.
+  void SetIcon(uint8_t* pixel_data, int width, int height) {
+    FlutterDesktopWindowSetIcon(window_, pixel_data, width, height);
+  }
+
+ private:
+  // Handle for interacting with the C API's window.
+  FlutterDesktopWindowRef window_;
```
Comment (chinmaygarde) [inline shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h]: I am unclear about how a window that is created but not consumed by FlutterDesktopRunWindowLoop is collected.

Comment (chinmaygarde) [reply shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h]: Maybe have a static method that takes a smart pointer to the FlutterWindow and consumes the same.

```
static void FlutterWindow::Run(std::unique_ptr<FlutterWindow>)
```

In the dtor, collect the window ref if the windows runloop was never run.

Comment (stuartmorgan) [reply shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h]: This is a preexisting issue with the C API; you can't not run the loop without leaking (see https://github.com/flutter/engine/blob/master/shell/platform/glfw/public/flutter_glfw.h#L50). There's no underlying method that can be called in the destructor at the moment, just as there wasn't when this code was part of `FlutterViewController`.

I was going to fix that when I reworked the runloop handling in the GLFW shell in general (there needs to be a way to do incremental loop runs, e.g., to allow having another window using a different runloop), but I can add just the stand-alone destruction method now.

Comment (chinmaygarde) [reply shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h]: sgtm

Comment (stuartmorgan) [reply shell/platform/glfw/client_wrapper/include/flutter/flutter_window.h]: Done. The GLFW API definitely still has structural issues that need to be resolved, but that's another step in the right direction.

**Pull request merged**